acceptance/github_asset_fetcher.go

```diff
@@ -0,0 +1,462 @@
+			return "", err
+		}
+
+		returnPath = targetDir
+	case "zip":
+		targetPath := filepath.Join(f.cacheDir, *desiredAsset.Name)
+		if err := f.downloadAndExtractZip(*desiredAsset.BrowserDownloadURL, targetPath); err != nil {
+			return "", err
+		}
+
+		returnPath = stripExtension(targetPath)
+	default:
+		targetPath := filepath.Join(f.cacheDir, *desiredAsset.Name)
+		if err := f.downloadAndSave(*desiredAsset.BrowserDownloadURL, targetPath); err != nil {
+			return "", err
+		}
+
+		returnPath = targetPath
+	}
+
+	err = f.writeCacheManifest(owner, repo, func(cache assetCache) {
+		existingAssets, found := cache[owner][repo].Assets[version]
+		if found {
+			cache[owner][repo].Assets[version] = append(existingAssets, returnPath)
+		}
+		cache[owner][repo].Assets[version] = []string{returnPath}
+	})
+	if err != nil {
+		f.logger.Warn(errors.Wrap(err, "writing cache").Error())
+	}
+	return returnPath, nil
+}
+
+func extractType(extract bool, assetName string) string {
+	if extract && strings.Contains(assetName, ".tgz") {
+		return "tgz"
+	}
+	if extract && strings.Contains(assetName, ".zip") {
+		return "zip"
+	}
+	return "none"
+}
+
+func (f *GithubAssetFetcher) FetchReleaseSource(owner, repo, version string) (string, error) {
+	if destDir, _ := f.cachedSource(owner, repo, version); destDir != "" {
+		fmt.Printf("found %s in cache for %s/%s %s\n", destDir, owner, repo, version)
+		return destDir, nil
+	}
+
+	release, _, err := f.githubClient.Repositories.GetReleaseByTag(f.ctx, owner, repo, version)
+	if err != nil {
+		return "", errors.Wrap(err, "getting release")
+	}
+
+	destDir := filepath.Join(f.cacheDir, strings.ReplaceAll(*release.Name, " ", "-")+"-source")
+	if err := os.MkdirAll(destDir, 0755); err != nil {
+		return "", errors.Wrapf(err, "creating directory %s", destDir)
+	}
+
+	if err := f.downloadAndExtractTgz(*release.TarballURL, destDir); err != nil {
+		return "", err
+	}
+
+	err = f.writeCacheManifest(owner, repo, func(cache assetCache) {
+		cache[owner][repo].Sources[version] = destDir
+	})
+	if err != nil {
+		f.logger.Warn(errors.Wrap(err, "writing cache").Error())
+	}
+	return destDir, nil
+}
+
+func (f *GithubAssetFetcher) FetchReleaseVersion(owner, repo string, n int) (string, error) {
+	if version, _ := f.cachedVersion(owner, repo, n); version != "" {
+		fmt.Printf("found %s in cache for %s/%s %d\n", version, owner, repo, n)
+		return version, nil
+	}
+
+	// get all release versions
+	releases, _, err := f.githubClient.Repositories.ListReleases(f.ctx, owner, repo, nil)
+	if err != nil {
+		return "", errors.Wrap(err, "listing releases")
+	}
+
+	// sort all release versions
+	versions := make([]*semver.Version, len(releases))
+	for i, release := range releases {
+		version, err := semver.NewVersion(*release.TagName)
+		if err != nil {
+			return "", errors.Wrap(err, "parsing semver")
+		}
+		versions[i] = version
+	}
+	sort.Sort(semver.Collection(versions))
+
+	latestVersion := versions[len(versions)-1]
```

Comment:
Shouldn't happen, but there doesn't seem to be a check if there are 0 releases.

Reply:
I will add this check.